crates/task-impls/src/quorum_proposal/handlers.rs

```diff
@@ -196,13 +203,41 @@ impl<TYPES: NodeType, V: Versions> ProposalDependencyHandle<TYPES, V> {
             )
             .await
             .wrap()
-            .context(warn!("Failed to construct marketplace block header"))?
+            .context(warn!("Failed to construct marketplace block header"))
+        };
+
+        let block_header = if version < V::Marketplace::VERSION {
+            get_legacy_block_header_fn.await?
+        } else if version < V::Epochs::VERSION {
+            get_marketplace_block_header_fn.await?
+        } else if self.consensus.read().await.is_high_qc_forming_eqc() {
+            tracing::info!("Reached end of epoch. Proposing the same block again to form an eQC.");
+            if let Some(leaf) = self
+                .consensus
+                .read()
+                .await
+                .saved_leaves()
+                .get(&high_qc.data().leaf_commit)
+            {
+                let block_header = leaf.block_header().clone();
+                tracing::debug!(
+                    "Proposing block no. {} to form the eQC.",
+                    block_header.block_number()
+                );
+                block_header
+            } else {
+                return Err(anyhow::anyhow!(
+                    "There is no leaf for the high QC. Consensus inconsistent!"
+                ));
+            }
+        } else {
+            get_marketplace_block_header_fn.await?
```

Comment:
we can't assume that `V::Epochs::VERSION >= V::Marketplace::VERSION` (I think actually the opposite) -- so I think this needs to be gated too

(I think the earlier logic where we copy the previous block header is fine, because we can expect it to be the same marketplace/legacy block header)

Reply:
Looks correct to me now

not sure how expensive `new_marketplace`/`new_legacy` are, but worth noting we are always calling them unnecessarily during epoch transitions here.

we can avoid that by making `let mut block_header = ` a future by removing the `await?`s, then overwriting the future (rather than the block header value) in line 216 with something like `async { Ok(parent_leaf.block_header().clone() }`. Then await the future only on line 224

(but if it's not an expensive call this isn't really important)